Bitsets3

wclodius2 opened the pull request:
Branch to implement bitset data types. Committed the following new source files:
src/stdlib_bitsets.f90
src/stdlib_bitset_64.f90
src/stdlib_bitset_large.f90

Added the following new test files:
src/tests/bitsets/CMakeLists.txt
src/tests/bitsets/Makefile.manual
src/tests/bitsets/test_stdlib_bitset_64.f90
src/tests/bitsets/test_stdlib_bitset_large.f90

Added the following new documentation file:
doc/specs/stdlib_bitssets.md

Modified the following compilation files:
src/CMakeLists.txt
src/Makefile.manual
src/tests/CMakeLists.txt
src/tests/Makefile.manual

jvdp1 commented on doc/specs/stdlib_bitsets.md:
```diff
@@ -0,0 +1,1975 @@
+            write(*,*) 'Initialization of logical(int32) succeeded.'
+        end if
+    end program demo_assignment
+
```
the ``` are missing
```suggestion
```
```

jvdp1 commented on doc/specs/stdlib_bitsets.md:
```diff
@@ -0,0 +1,1975 @@
+            write(*,*) 'SET1 has the proper size.'
+        if ( set1 % all() ) write(*,*) 'SET1 has the proper values.'
+    end program demo_extract
+
```
The ``` are missing

jvdp1 commented on doc/specs/stdlib_bitsets.md:
```diff
@@ -0,0 +1,1975 @@
+        call set0 % flip(0,164)
+        if ( set0 % all() ) write(*,*) 'All bits are flipped.'
+    end program demo_flip
+
```
the ``` are missing

jvdp1 reviewed (commented):
Thank you @wclodius2 for this PR. I will try to review it this weekend or next week.

jvdp1 commented on src/tests/Makefile.manual:
```diff
@@ -10,6 +11,7 @@ all:
 
 test:
 	$(MAKE) -f Makefile.manual --directory=ascii test
+	$(MAKE) -f Makefile.manual --directoru=bitsets test
```
```suggestion
	$(MAKE) -f Makefile.manual --directory=bitsets test
```
This should allow the Github actions to be successful

14NGiestas reviewed (commented):
There's a misspelling of the word 'occurred' in this file: "occured" (line 1092)
EDIT: [this file](https://github.com/wclodius2/stdlib/blob/bitsets3/doc/specs/stdlib_bitsets.md)

jvdp1 commented on doc/specs/stdlib_bitsets.md:
```diff
@@ -0,0 +1,1990 @@
+
+* `success` - no problem found
+
+* `write_failure` - a failure occured in a write statement.
```
```suggestion
* `write_failure` - a failure occurred in a write statement.
```

wclodius2 commented:
Looks good.

> On Oct 4, 2020, at 11:49 AM, Jeremie Vandenplas <notifications@github.com> wrote:
> 
> 
> @jvdp1 commented on this pull request.
> 
> In src/tests/Makefile.manual <https://github.com/fortran-lang/stdlib/pull/239#discussion_r499271200>:
> 
> > @@ -10,6 +11,7 @@ all:
>  
>  test:
>  	$(MAKE) -f Makefile.manual --directory=ascii test
> +	$(MAKE) -f Makefile.manual --directoru=bitsets test
> ⬇️ Suggested change
> -	$(MAKE) -f Makefile.manual --directoru=bitsets test
> +	$(MAKE) -f Makefile.manual --directory=bitsets test
> This should allow the Github actions to be successful
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/fortran-lang/stdlib/pull/239#pullrequestreview-501644579>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/APTQDOWITSG6CMZ2MNAX2VLSJCYSBANCNFSM4R6QFU5A>.
>

14NGiestas commented:
We should really discuss the error handling further in #219 . The 'go to' usage in [this function](https://github.com/wclodius2/stdlib/blob/bitsets3/src/stdlib_bitset_large.f90#L960)... I'm not a fan xD. I think only one labeled part is needed to achieve the 'return with status' vs 'stop program with error message' desired behavior, with the aid of a error handling module, we could set the message and a flag to switch between the two possible actions, reusing the code. I will bring some working code snippet soon (I hope).
Also, in the assigment functions I wonder if it is possible to use the template system (as in the statistics module)?

wclodius2 commented:
What is xD? The error handling has several aspects that could be commented on:

1. Rather than intermingling the error handling with the main logic I jump to error handling at the end of the procedure. This increases the amount of code slightly, and can obscure what the error handling is doing, but I think makes the main logic clearer.

2. I am handling some "errors" that possibly shouldn't be handled: memory allocation problems and user errors. With virtual memory, its not clear to me that allocation can ever fail, and with lazy allocation on Linux such failure will not occur on the allocate statement.. While not important in the procedure you call out, I also attempt to provide a fallback for some user errors where simple failure might be best.

3. Since we lack an error handling procedure I am relying explicitly on an if branch with `error stop`. An error handling procedure could incorporate the if branch and slightly simplify the logic by incorporating both a message and a status argument. There are however a lot of issues to be dealt with in developing an error handling module.

    * Should it be a separate module or part of the logger?

    * Should it include a set of error codes?

    * What should the arguments be to the main error handling procedure?
        - message
        - module and procedure names
        - `status`/`stat` flag
        - error code

    * How should processing be terminated, an `error stop`: with an F2018 runtime determined stop code, with an error code determined compile time stop code, generic compile time stop code?

4. Including the module and procedure names in the `error stop` makes the stop code longer and harder to read

5. Making the stop code message a character string parameter would shorten the error handling code, at the expense of cluttering up other portions of the code.

It is of course possible to use fypp preprocessing to generate templates for the assignments to and from logical arrays. The assignment procedures are few enough and simple enough that I thought it best to avoid a dependence on the preprocessor, but that is a matter of taste. The addition of more assignments, say to and from integers would probably change my opinion.

14NGiestas commented:
> What is xD?

ASCII version of 😆smiling face with closed eyes

1-5. Yes, the current way increases the amount of code, however the main logic can be still clear if we do something like that:

```fortran
    subroutine fail_behavior(..., status)
        ! doing a lot of stuff here
        ! ...
        ! case X: found some fatal error
            error_handler % status = ERROR_CODE_TO_X
            error_handler % message = "Oh no! X happened"
            go to 404
        ! case Y: found other general error
            error_handler % status = ERROR_CODE_TO_Y
            error_handler % message = "Oh no! Y happened"
            go to 404

    404 if (present(status)) then
            status = error_handler % status
            return
        else
            error stop error_handler % message
        end if
    end subroutine
``` 
This is a way we can avoid the code duplication, reusing the go to branching. Note that error_handler should be a object that couples the information about what happened (not reflecting in any way the API I got in mind for a final version).
The error handling should be a separate module with some default general usage errors and a abstract one, that would allow it to be extended when needed by other modules, this way they can provide their own specific codes constants and messages.

> It is of course possible to use fypp preprocessing to generate templates for the assignments to and from logical arrays. The assignment procedures are few enough and simple enough that I thought it best to avoid a dependence on the preprocessor, but that is a matter of taste. The addition of more assignments, say to and from integers would probably change my opinion.

IMHO, we should use the template system every time that not using it leads to code duplication, since it gets easier to modify and extend the code in the future. (DRY principle).

wclodius2 commented:
FWIW for the error handling what I thought you were proposing was something like

    subroutine fail_behavior(..., status)
        ! doing a lot of stuff here
        ! ...
        ! case X: found some fatal error
            call error_handler(message="Oh no! X happened", error=error_code_to_x, status=status)
            return
        ! case Y: found other general error
            call error_handler(message="Oh no! Y happened", error=error_code_to_y, status=status)
            return
    end subroutine

where `error_handler` assigns `error` to `status` if `status` is `present` and stops processing with a message to `error_unit` otherwise.

I can change the code to use `fypp`, but it probably requires starting a new branch. I can see how to add or modify files in the current branch, but not how to rename or remove existing files, which I need to change the `*.f90` files to `*.fypp`.

14NGiestas commented:
> FWIW for the error handling what I thought you were proposing was something like (...)
> where `error_handler` assigns `error` to `status` if `status` is `present` and stops processing with a message to `error_unit` otherwise.

Indeed, actually this is even better because we don't need retype the branching logic with "go to" statements in every function that may fail and is doable [(working example here)](https://gist.github.com/14NGiestas/8ebf6753e6d4680ecd044faa2c2bed85). To avoid having any boilerplate along the function's main logic, the module that have a specific need should build a function that calls this base function `error_handler`, Ex:
```fortran
subroutine value_error(expected, got, status)
     ....
     write(message, '("value_error: Expected ",G0,", but got ",G0," instead.")') expected, got
     call error_handler(message, ...) 
end subroutine
```

> I can change the code to use `fypp`, but it probably requires starting a new branch. I can see how to add or modify files in the current branch, but not how to rename or remove existing files, which I need to change the `*.f90` files to `*.fypp`.

You can clone your repo [here](https://github.com/wclodius2/stdlib/tree/bitsets3), modify it locally and push the commits (It will probably sync with the open PR)

jvdp1 commented:
> I can see how to add or modify files in the current branch, but not how to rename or remove existing files, which I need to change the `*.f90` files to `*.fypp`.

You can rename and remove a file locally with your OS, and then add/remove, commit, and push with git.

wclodius2 commented:
When I try

    git clone https://github.com/wclodius2/stdlib/tree/bitsets3 my-bitsets

I get the error message

    Cloning into 'my-bitsets'...
    fatal: repository 'https://github.com/wclodius2/stdlib/tree/bitsets3/' not found

Is there something wrong with my syntax/repository name?

14NGiestas commented:
This is the wrong URL, please do the following:
```
git clone https://github.com/wclodius2/stdlib.git my-bitsets
cd my-bitsets
git checkout bitsets3
```

14NGiestas commented on src/stdlib_bitset_large.fypp:
```diff
@@ -1396,17 +1342,11 @@ contains
-999     if ( present(status) ) then
-            status = write_failure
+        if (ierr /= 0) then
+            call error_handler( 'Failure on a WRiTE statement for UNIT.', &
```
```suggestion
            call error_handler( 'Failure on a WRITE statement for UNIT.', &
```

14NGiestas commented on src/stdlib_bitset_large.fypp:
```diff
@@ -807,31 +785,29 @@ contains
-            error stop module_name // ' % ' // procedure // ' had a ' // &
-                'failure in the write to UNIT.'
-        end if
+999     call error_handler( 'Failure on a WRiTE statement for UNIT.', &
```
```suggestion
999     call error_handler( 'Failure on a WRITE statement for UNIT.', &
```

wclodius2 commented:
I have modified, committed and pushed the code making the following changes:
1. Changed  the main file names from stdlib_bitset*.f90 to stdlib_bitset*.fypp
2. Used the preprocessor to autogenerate the assignment defs for logical arrays
3. Added a subroutine error_handler to stdlib_bitsets.fypp
4. Used the subroutine to handle errors in stdlib_bitset_64.fypp and stdlib_bitset_large.fypp
5. Used bulleted lists to document the use of error codes in stdlib_bitset*.fypp and stdlib_bitsets.md
6. Added the error code char_string_too_large_error
7. After pushing realized I forgot to check for trailing blanks. Removed trailing blanks and pushed again.

wclodius2 commented:
Failed to run tests on the previous update.  I had accidentally removed a return statement causing failure of test_stdlib_bitset_64. Fixed and pushed.

14NGiestas commented:
Thank you, @wclodius2!

14NGiestas commented on src/stdlib_bitset_64.fypp:
```diff
@@ -0,0 +1,1145 @@
+            return
+        end if
+        self % num_bits = bits
+        do bit = 1, bits
```
I get this warning compiling it:
```
/home/ian/stdlib/src/stdlib_bitset_64.f90:437:19:

  437 |         do bit = 1, bits
      |                   1
Warning: Possible change of value in conversion from INTEGER(8) to INTEGER(4) at (1) [-Wconversion]
```
Maybe changing to this:
```suggestion
        do bit = 1_bits_kind, bits
```
Or modifying the kind of the variable `bit` to other than `bits_kind` since it's a simple iterator

wclodius2 replied:
If you check on the use of bits, it is declared int64 to always ensure that an int32 for bit is adequate. The warning is not for the 1 in `do bit = 1, bits`, but rather for the conversion of `bits` to `bit`. The following tests ensure that int32 is adequate.

        bits = len(string, kind=int64)
        if ( bits > 64 ) then
 
in stdlib_bitset_64.fypp or

        bits = len(string, kind=int64)
        if ( bits > huge(0_bits_kind) ) then

in stdlib_bitset_large.fypp. So the warning is spurious. I can, if you want, shut off the warning by making bit an int64.

14NGiestas replied:
That's correct, it's `bit` vs `bits` conversion that raises this warning. Is there any reason to `bit` be of kind `bits_kind`, in this function? If not, I think it should be a `int64` as well.

14NGiestas commented on src/stdlib_bitset_large.fypp:
```diff
@@ -0,0 +1,1372 @@
+            if ( status /= success ) return
+        end if
+
+        do bit = 1_bits_kind, bits
```
Here there's a similar warning, this time the other way around:
```
/home/ian/stdlib/src/stdlib_bitset_large.f90:562:29:

  562 |         do bit = 1_bits_kind, bits
      |                             1
Warning: Possible change of value in conversion from INTEGER(8) to INTEGER(4) at (1) [-Wconversion]
```
I believe `bits` and `bit` should have the same kind, other than `bits_kind`, and since they are iterators there's no problem using `int64` as kind here.

wclodius2 commented:
I have made some aesthetic changes in the source code: deleting a couple of unnecessary returns, and regularizing the code fences in the FORD documentation of stdlib_bitsets.fypp. I now plan to take a break of a few days from edits of the source code, unless they are specifically requested.

jvdp1 commented:
> I have made some aesthetic changes in the source code: deleting a couple of unnecessary returns, and regularizing the code fences in the FORD documentation of stdlib_bitsets.fypp. I now plan to take a break of a few days from edits of the source code, unless they are specifically requested.

Thank you @wclodius2 . I will try to review it next week,

jvdp1 commented on doc/specs/stdlib_bitsets.md:
```diff
@@ -0,0 +1,2006 @@
+
+`set2`: shall be a scalar expression of the same type as `set1`. It is
+an `intent(in)` argument. Note that it should also have the same
+number of bits as `set1` otherwise the result is undefined.
```
```suggestion
number of bits as `set1`, otherwise the result is undefined.
```

jvdp1 commented on doc/specs/stdlib_bitsets.md:
```diff
@@ -0,0 +1,2006 @@
+`self`: shall be a scalar expression of class `bitset_type`. It is an
+`intent(in)` argument.
+
+#### Result character
+
+The result is a default logical scalar.
+
+#### Result value
+
+The result is `.true.` if any bits in `self` are set, otherwise it
+is `.false.`.
```
These 2 sections could be merged. Currently it is confusing.
```suggestion
#### Result value

The result is a default logical scalar. The result is `.true.` if any bits in `self` are set, otherwise it
is `.false.`.
```

wclodius2 replied:
OK

jvdp1 commented on doc/specs/stdlib_bitsets.md:
```diff
@@ -0,0 +1,2006 @@
+`self`: shall be a scalar expression of class `bitset_type`. It is an
+`intent(in)` argument.
+
+#### Result character
+
+The result is an integer scalar of kind `bits_kind`.
+
+#### Result value
+
+The result is the number of bits that are set in `self`.
+
```
```suggestion
#### Result value

The result is an integer scalar of kind `bits_kind`, equal to the number of bits that are set in `self`.

```

jvdp1 commented on doc/specs/stdlib_bitsets.md:
```diff
@@ -0,0 +1,2006 @@
+* if `start_pos` and `end_pos` are present with `end_pos < start_pos`
+`self` is unmodified.
+
+Note: Positions outside the range 0 to `BITS(SET) -1` are ignored.
```
What happens with the range  -9 to 'bits(set)+10`?
```suggestion
Note: Positions outside the range 0 to `bits(set) -1` are ignored.
```

jvdp1 replied:
Reminder for myself: Are there some tests with a range outside 0-`bits(set)-`?

jvdp1 commented on doc/specs/stdlib_bitsets.md:
```diff
@@ -0,0 +1,2006 @@
+
+#### Class
+
+Subroutine
```
Why is this a subroutine instead a function, e.g.:
`new = extract(old, start_pos,...)`?

Note: I am fine (and even prefer it) as a subroutine. It is mainly to document the PR.

wclodius2 replied:
The use of a subroutine:
1. Makes it easier to enforce that `new` must be the same type as `old`
2. Allows me to return a status flag, which is useful now, and necessary if I ever want to allow `new` and `old` to be mixed types

jvdp1 commented on doc/specs/stdlib_bitsets.md:
```diff
@@ -0,0 +1,2006 @@
+
+`bits` (optional): shall be a scalar integer expression of kind
+`bits_kind`. It is an `intent(in)` argument that if present
+specifies the number of bits in `set`. A negative value or a value
+greater than 64 if `self` is of type `bitset_64` is an error.
```
```suggestion
specifies the number of bits in `set`. A negative value, or a value
greater than 64 if `self` is of type `bitset_64`, is an error.
```

jvdp1 commented on doc/specs/stdlib_bitsets.md:
```diff
@@ -0,0 +1,2006 @@
+#### Description
+
+Reads a bitset from its binary representation in an unformatted
+file.
```
Is it not worthwhile to mention the type of access used for unformatted files?

wclodius2 replied:
I had been assuming sequential access, but it doesn't matter as long as the read is in the right position in the file. To ensure that you probably want to use the same access in the read as was used in the write.

jvdp1 commented on doc/specs/stdlib_bitsets.md:
```diff
@@ -0,0 +1,2006 @@
+
+`status` (optional): shall be a scalar default integer variable. If
+present its value shall be of one of the error codes defined in this
+module. IF absent and it would have had a value other than `success`
```
```suggestion
module. If absent and it would have had a value other than `success`
```

jvdp1 commented on doc/specs/stdlib_bitsets.md:
```diff
@@ -0,0 +1,2006 @@
+        call set2 % from_string( bits_33 )
+        open( newunit=unit, file='test.bin', status='replace', &
+            form='unformatted', action='write' )
+        call set2 % output(unit)
```
Maybe I missed it, but would it be possible to use:
```fortran
 call set2 % output('test.bin')
```
and similarly for `input`?

jvdp1 commented on doc/specs/stdlib_bitsets.md:
```diff
@@ -0,0 +1,2006 @@
+`self`: shall be a scalar expression of class `bitset_type`. It is an
+  `intent(in)` argument.
+
+#### Result character
+
+The result is a default logical scalar.
+
+#### Result value
+
+The result is `.true.` if no bits in `self` are set, otherwise it is
+`.false.`.
```
```suggestion
#### Result value

The result is a default logical scalar.
The result is `.true.` if no bits in `self` are set, otherwise it is
`.false.`.
```

jvdp1 commented on doc/specs/stdlib_bitsets.md:
```diff
@@ -0,0 +1,2006 @@
+
+#### Syntax
+
+`result = self % [[bitset_type(class):not(bound)]] ()`
```
```suggestion
`call self % [[bitset_type(class):not(bound)]] ()`
```

jvdp1 commented on doc/specs/stdlib_bitsets.md:
```diff
@@ -0,0 +1,2006 @@
+
+Replaces the original bits of `set1` with the bitwise `or` of those
+bits with the bits of `set2`. Note `set1` and `set2` must have the
+samee number of bits, otherwise the result is undefined.
```
```suggestion
same number of bits, otherwise the result is undefined.
```

jvdp1 replied:
How is `undefined` defined? `NaN? (I didn't look to the code yet, but I think it would be good to define "undefined" in the specs.)

jvdp1 commented on doc/specs/stdlib_bitsets.md:
```diff
@@ -0,0 +1,2006 @@
+ corresponding bits in `set2`.
+
+`set2`: shall be a scalar expression of the same type as `set1`. It is
+an `intent(in)` argument. Note `bits(set2)` must equal `bits(set1)`
```
```suggestion
an `intent(in)` argument. Note that `bits(set2)` must equal `bits(set1)`
```

jvdp1 commented on doc/specs/stdlib_bitsets.md:
```diff
@@ -0,0 +1,2006 @@
+
+#### Description
+
+Writes a binary representation of a bitset to an unformatted file.
```
What kind of access is used? It might be important to mention it because the user must open the file himself.

jvdp1 commented on doc/specs/stdlib_bitsets.md:
```diff
@@ -0,0 +1,2006 @@
+
+#### Arguments
+
+`self`: shall be a scalar `bitset_64` or `bitset_large` variable. It
```
to what are set the bits? All `0`? All `1`? Uninitialized?

jvdp1 commented on doc/specs/stdlib_bitsets.md:
```diff
@@ -0,0 +1,2006 @@
+
+#### Result value
+
+The result is `.true.` if the bit at `pos`, in `self` is set,
```
```suggestion
The result is `.true.` if the bit at `pos` in `self` is set,
```

jvdp1 commented on doc/specs/stdlib_bitsets.md:
```diff
@@ -0,0 +1,2006 @@
+#### Result value
+
+The result is `.true.` if the bit at `pos`, in `self` is set,
+otherwise it is `.false.`. If `pos` is outside the range
+`0... bits(self)-1` the result is `.false.`.
+
```
" If `pos` is outside the range ... `.false.`.": does this make sense? As a user, I would prefer that the program stops, or at least, returns something corresponding to `NaN`

jvdp1 commented on doc/specs/stdlib_bitsets.md:
```diff
@@ -0,0 +1,2006 @@
+
+#### Syntax
+
+`call self % [[bitset_type(class):to_string(bound)]](string[, status])
```
```suggestion
`call self % [[bitset_type(class):to_string(bound)]](string[, status])`
```

jvdp1 commented on doc/specs/stdlib_bitsets.md:
```diff
@@ -0,0 +1,2006 @@
+
+#### Status
+
+Experimeental
```
```suggestion
Experimental
```

jvdp1 commented on doc/specs/stdlib_bitsets.md:
```diff
@@ -0,0 +1,2006 @@
+`pos`: shall be a scalar integer expression of kind `bits_kind`. It is
+an `intent(in)` argument.
+
+#### Result character
+
+The result is a default logical scalar.
+
+#### Result value
+
+The result is one if the bit at `pos` in `self` is set, otherwise it
+is zero. If `pos` is outside the range `0... bits(set)-1` the result
+is zero.
+
```
```suggestion
#### Result value

The result is a default logical scalar.
The result is 1 if the bit at `pos` in `self` is set, otherwise it
is 0. If `pos` is outside the range `0... bits(set)-1` the result
is 0.

```

Again, does it make sense to return 0 if pos is outside the range?

jvdp1 commented on doc/specs/stdlib_bitsets.md:
```diff
@@ -0,0 +1,2006 @@
+
+#### Syntax
+
+`Result = set1 [[stdlib_bitsets(module):==(interface)]] set2
```
```suggestion
`result = set1 [[stdlib_bitsets(module):==(interface)]] set2
```

jvdp1 commented on doc/specs/stdlib_bitsets.md:
```diff
@@ -0,0 +1,2006 @@
+will have the same number of bits as `set1`. It is an `intent(in)`
+argument.
+
+#### Result character
+
+The result is a default logical scalar.
+
+#### Result value
+
+The result is `.true.` if the bits in both bitsets are set
+  to the same value, otherwise the result is `.FALSE.`.
```
```suggestion
#### Result value

The result is a default logical scalar.
The result is `.true.` if the bits in both bitsets are set
  to the same value, otherwise the result is `.false.`.
```

jvdp1 commented on doc/specs/stdlib_bitsets.md:
```diff
@@ -0,0 +1,2006 @@
+
+The result is `.true.` if the bits in `set1` and `set2` are the same
+or the highest order different bit is set to 1 in `set1` and to 0 in
+`set2`, `.false.`.  otherwise.
```
```suggestion
`set2`, `.false.`  otherwise.
```

jvdp1 commented on doc/specs/stdlib_bitsets.md:
```diff
@@ -0,0 +1,2006 @@
+
+Returns `.true.` if the bits in `set1` and `set2` differ and the
+highest order different bit is set to 1 in `set1` and to 0 in `set2`,
+`.false.`.  otherwise. The sets must be the same size otherwise the
```
```suggestion
`.false.`  otherwise. The sets must be the same size otherwise the
```

jvdp1 commented on doc/specs/stdlib_bitsets.md:
```diff
@@ -0,0 +1,2006 @@
+
+#### Syntax
+
+`Result = set1 [[stdlib_bitsets(module):>(interface)]] set2`
```
```suggestion
`result = set1 [[stdlib_bitsets(module):>(interface)]] set2`
```

jvdp1 commented on doc/specs/stdlib_bitsets.md:
```diff
@@ -0,0 +1,2006 @@
+
+Returns `.true.` if the bits in `set1` and `set2` are the same or the
+highest order different bit is set to 0 in `set1` and to 1 in `set2`,
+`.false.`.  otherwise. The sets must be the same size otherwise the
```
```suggestion
`.false.`  otherwise. The sets must be the same size otherwise the
```

jvdp1 commented on doc/specs/stdlib_bitsets.md:
```diff
@@ -0,0 +1,2006 @@
+
+#### Syntax
+
+`Result = set1 [[stdlib_bitsets(module):<=(interface)]] set2`
```
```suggestion
`result = set1 [[stdlib_bitsets(module):<=(interface)]] set2`
```

jvdp1 commented on doc/specs/stdlib_bitsets.md:
```diff
@@ -0,0 +1,2006 @@
+
+The result is `.true.` if the bits in `set1` and `set2` are the same
+or the highest order different bit is set to 0 in `set1` and to 1 in
+`set2`, `.false.`.  otherwise.
```
```suggestion
`set2`, `.false.`  otherwise.
```

jvdp1 commented on doc/specs/stdlib_bitsets.md:
```diff
@@ -0,0 +1,2006 @@
+
+Returns `.true.` if the bits in `set1` and `set2` differ and the
+highest order different bit is set to 0 in `set1` and to 1 in `set2`,
+`.false.`.  otherwise. The sets must be the same size otherwise the
```
```suggestion
`.false.`  otherwise. The sets must be the same size otherwise the
```

jvdp1 commented on doc/specs/stdlib_bitsets.md:
```diff
@@ -0,0 +1,2006 @@
+
+#### Syntax
+
+`Result = set1 [[stdlib_bitsets(module):<(interface)]] set2`
```
```suggestion
`result = set1 [[stdlib_bitsets(module):<(interface)]] set2`
```

jvdp1 commented on src/CMakeLists.txt:
```diff
@@ -2,6 +2,9 @@
 
 # Create a list of the files to be preprocessed
 set(fppFiles
+    stdlib_bitset_64.fypp
+    stdlib_bitset_large.fypp
```
I suggest to call these 2 files as follows:
```suggestion
    stdlib_bitsets_64.fypp
    stdlib_bitsets_large.fypp
```

wclodius2 replied:
What are your reasons for changing the file names? Changing them affects five files, the two files whose names are changed, them CMakeLists.txt file, the Makefile.manual file, and the markdown document.

jvdp1 replied:
So they would have the same prefix ("stdlib_bitsets") as the file stdlib_bitsets.fypp. I always find such things easier to edit (e.g. I don't need to think if a "s" is needed or not). If others have no objections against the current names, it is also fine for me.

jvdp1 reviewed (commented):
Thank you @wclodius2 for this work.
I just reviewed the specs and it looks good to me. I  still need to review the source code and play with it to get more opinions.

wclodius2 commented on doc/specs/stdlib_bitsets.md:
```diff
@@ -0,0 +1,2006 @@
+literal that can be read to and from strings and formatted files, a
+simple character string literal that can be read to and from strings,
+assignments, procedures, methods, and operators. Note that the module
+assumes two's complement integers, but all current Fortran 95+ processors use such integers.
```
It is necessary if users are concerned about the portability of their code. There are still some Unisys machines in operation, and they used one's complement, but never had a Fortran 90+ compiler. There may also be some microcontrollers with one's complement or signed magnitude implementations.

jvdp1 replied:
Thank you. I 'll remove my suggestion.

wclodius2 commented on doc/specs/stdlib_bitsets.md:
```diff
@@ -0,0 +1,2006 @@
+for each element of the sequence, where `test` is `.true.` if the value
+is 1 and `.false.` otherwise. The number of such values in an entity
+of that type is to be termed, `bits`. The bits are ordered in terms of
+position, that, in turn, is indexed from 0 to `bits-1`. `bitset_type` is
```
It can be used to declare a class variable (`class(bitset_type)`) that can be used to represent either bitset_64 or bitset_large types.

wclodius2 commented on doc/specs/stdlib_bitsets.md:
````diff
@@ -0,0 +1,2006 @@
+    end program demo_assignment
+```
+
+### Table of the `bitset_type` methods
````
OK

wclodius2 commented on doc/specs/stdlib_bitsets.md:
```diff
@@ -0,0 +1,2006 @@
+number of bits otherwise the results are undefined. These operators
+are summarized in the following table:
+
+|Operator|Description|
```
My understanding is that ==, etc. are synonyms for .EQ., etc. and defining one automatically defines the other. If I remember correctly gfortran objected when I tried to define both pairs separately on another project. I could list them in this table for thoroughness.

jvdp1 replied:
Ok. If defining one automatically defines the other, then I think it would be good to add the other ones in the table too.

wclodius2 commented on doc/specs/stdlib_bitsets.md:
```diff
@@ -0,0 +1,2006 @@
+#### Argument
+
+`self`: shall be a scalar expression of class `bitset_type`. It is an
+`intent(in)` argument.
```
The form I use parallels that used in the standard. Your form would be awkward for optional arguments.

jvdp1 replied:
Ok. I 'll remove my comment. Thank you for the explanation.

wclodius2 commented on doc/specs/stdlib_bitsets.md:
```diff
@@ -0,0 +1,2006 @@
+
+Sets the bits in `set1` to the bitwise `and` of the original bits in
+`set1` and `set2`. Note that `set1` and `set2` must have the same
+number of bits, otherwise the result is undefined.
```
NaN is solely for floating point (real) numbers. The value of the individual bits in the result is undefined. In practice if `set1` has more bytes than `set2`, then you can have a segmentation fault. If `set2` has more bits than `set1` then bits in `set1` beyond `bits(set1)-1` may be modified, which can have an effect if `set1` is used as the second operand in a binary operation. If `set1` has more bits than `set2`, but the same number of bytes, then the bits beyond `bits(set2)-1` will be used in the operation.

jvdp1 replied:
thank you for the explanation. IMO it would be good to add these explanations to the specs. Let see what others think about it.

jvdp1 commented on doc/specs/stdlib_bitsets.md:
```diff
@@ -0,0 +1,2002 @@
+is 1 and `.false.` otherwise. The number of such values in an entity
+of that type is to be termed, `bits`. The bits are ordered in terms of
+position, that, in turn, is indexed from 0 to `bits-1`. `bitset_type` is
+not used in source code. The syntax for using the types are:
```
```suggestion
not used in source code of `stdlib`, but is publicaly avaliable fo users to represent either `bitset_64` or `bitset_large`.
The syntax for using the types are:
```
Related to [this comment](https://github.com/fortran-lang/stdlib/pull/239/files#r504891872)

wclodius2 replied:
I don't think this is correct. I have changed the text to

used only as a `class` to define entities that can be either a `bitset_64` or
a `bitset_large`. The syntax for using the types are:

Committed and pushed.

wclodius2 commented:
@jvdp1 I have just committed edits to stdlib_bitsets.md based on most of your suggestions in your first posting. I will now rename the fypp files, change the names of their submodules to match the file names, and change the corresponding CMakeLists.txt and Makefile.manual files

jvdp1 commented on src/stdlib_bitsets.fypp:
```diff
@@ -0,0 +1,2126 @@
+    integer, parameter ::                    &
+        block_size  = bit_size(0_block_kind)
+
+    integer(block_kind), private, parameter :: all_zeros  = 0_block_kind
+    integer(block_kind), private, parameter :: all_ones   = not(all_zeros)
+
+    character(*), parameter, private :: module_name = "STDLIB_BITSETS"
+    integer, parameter, private :: &
```
```suggestion
    integer(block_kind),  parameter :: all_zeros  = 0_block_kind
    integer(block_kind),  parameter :: all_ones   = not(all_zeros)

    character(*), parameter :: module_name = "STDLIB_BITSETS"
    integer, parameter :: &
```
`private` is not needed because the default in the module is `private`

jvdp1 commented on src/stdlib_bitsets.fypp:
```diff
@@ -0,0 +1,2126 @@
+!!
+!! Parent type for bitset_64 and bitset_large
+        private
+        integer(bits_kind) :: num_bits
```
Currently `bits_kind` is equal to `int32`.
I am interested by `stdlib_bitsets` because I have a real case for such a module. However, limiting the number of bits by using `int32` would not allow me to use `stdlib_bitsets` in my applications. The max of int32 is easily reached. What would be the costs to use `int64` instead of `int32`?

wclodius2 replied:
You have a need for greater than 2**31-1 bits in a single variable, as opposed to blocks of 64 bits? I will have to check the code to be certain I was consistent in using bits_kind, but it was intended that it could be changed if desired by the user. The variable will be greater than (2**31-1)/8 bytes so the caches will be too small to hold a single variable so unless you maintain locality of reference (move through the array sequentially) memory traffic might be high.

I have checked the source code. There were problems with stdlib_bitsets_large.fypp,, primarily in not requiring size to consistently return a `bits_kind` integer `kind`, but also a string `len` and the subroutine digit_count. I also made sure the do loop bounds were consistently `bits_kind`. I think I have fixed all the problems, and the typos you found in stdlib_bitsets.fypp, and have committed  and pushed the revised code.

wclodius2 replied:
My fix for the subroutine digit_count works for gfortran, but not for ifort where it reports errors at compile time. It doesn't like case values out of range of the select value's kind. gfortran only gives warnings. I have ideas for a fix, but don't have time today.

wclodius2 replied:
I fixed the ifort problem, replacing the select case construct with a do loop. I have committed and pushed the resulting code. Now the code is passing all three MacOS builds and failing all three Ubuntu builds with the message:

Error: /home/runner/work/stdlib/stdlib/build is not a directory
Error: Process completed with exit code 1.

I only have a Mac to test on, and don't know the configuration of the Ubuntu systems, so I am at a loss for how to fix things.

jvdp1 replied:
> I fixed the ifort problem, replacing the select case construct with a do loop. I have committed and pushed the resulting code. Now the code is passing all three MacOS builds and failing all three Ubuntu builds with the message:
> 
> Error: /home/runner/work/stdlib/stdlib/build is not a directory
> Error: Process completed with exit code 1.
> 
> I only have a Mac to test on, and don't know the configuration of the Ubuntu systems, so I am at a loss for how to fix things.

Thank you. I will check that later on my Linux computer.

wclodius2 replied:
I suspect the problem was that I hadn't updated Makefile.manual to reflect the use of stdlib_kinds.f90. Upon updating that, and the other codes so that they compile and run with `bits_kind=int64`, the committed and pushed codes do not report any errors in the past two hours. Usually it only takes a few minutes for them to report a problem.

jvdp1 commented on src/stdlib_bitsets.fypp:
```diff
@@ -0,0 +1,2126 @@
+!! * `success` -  if no problems were found,
+!! * `alloc_fault` - if memory allocation failed
+!! * `array_size_invalid_error` - if `bits` is either negative or larger
+!!   than 64 with `self` of class `bitset_64`, or
```
This sentence is not finished, right?

wclodius2 replied:
As near as I could tell the dangling ", or" was inappropriate and my revisions have replaced it with a "."

jvdp1 commented on src/stdlib_bitsets.fypp:
```diff
@@ -0,0 +1,2126 @@
+        elemental module subroutine clear_bit_64(self, pos)
+!! Version: experimental
+!!
+!! Sets to zero the bit at `pos` position in `self`. If 'pos` is less than
```
```suggestion
!! Sets to zero the bit at `pos` position in `self`. If `pos` is less than
```

jvdp1 commented on src/stdlib_bitsets.fypp:
```diff
@@ -0,0 +1,2126 @@
+!! Version: experimental
+!!
+!! Sets the bits in `set1` to the bitwise `and` of the original bits in `set1`
+!! and `set2`. The sets mmust have the same number of bits
```
```suggestion
!! and `set2`. The sets must have the same number of bits
```

jvdp1 commented on src/stdlib_bitsets.fypp:
```diff
@@ -0,0 +1,2126 @@
+!! Version: experimental
+!!
+!! Creates a new bitset, `new`, from a range, `start_pos` to `stop_pos`, in
+!! bitset `old`. If 'start_pos` is greater than `stop_pos` the new bitset is
```
```suggestion
!! bitset `old`. If `start_pos` is greater than `stop_pos` the new bitset is
```

wclodius2 commented on doc/specs/stdlib_bitsets.md:
```diff
@@ -0,0 +1,2006 @@
+the integer kind value for indexing bits and reporting counts of
+bits. The other constants that are error codes are summarized below:
+
+|Error Code|Summary|
```
Wait until the reviewers have seen what I did in the source code.

jvdp1 commented on src/stdlib_bitsets.fypp:
```diff
@@ -0,0 +1,2126 @@
+        elemental module function eqv_large(set1, set2) result(eqv)
+!! Version: experimental
+!!
+!! Returns `.true`. if all bits in `set1` and `set2` have the same value,
```
```suggestion
!! Returns `.true` if all bits in `set1` and `set2` have the same value,
```

jvdp1 commented on src/stdlib_bitsets.fypp:
```diff
@@ -0,0 +1,2126 @@
+!! Version: experimental
+!!
+!! Returns `.true.` if the bits in `set1` and `set2` differ and the
+!! highest order different bit is set to 1 in `set1` and to 0 in `set2`.
```
```suggestion
!! highest order different bit is set to 1 in `set1` and to 0 in `set2`,
```

jvdp1 reviewed (commented):
Here are some other small comments.

wclodius2 commented on src/stdlib_bitsets.fypp:
```diff
@@ -0,0 +1,2126 @@
+!! having the sign bit set.
+
+    use, intrinsic ::            &
+        iso_fortran_env, only:   &
```
Force of habit. I will change it.

jvdp1 reviewed (commented):
Some typos (I will commit most of them directly).

wclodius2 commented on src/stdlib_bitsets_64.fypp:
```diff
@@ -0,0 +1,1120 @@
+                bits = bits*10 + iachar( string(pos:pos) ) - ia0
+                if ( bits < 0 ) go to 996
+            case(iachar('b'), iachar('B'))
+                go to 100
```
For some reason I missed this and the following post. I will replace the `go to 100` with a simple `exit` for the moment. Let me know if that is unsatisfactory for some reason.

wclodius2 commented on src/stdlib_bitsets_64.fypp:
```diff
@@ -0,0 +1,1120 @@
+
+        end do
+
+100     if ( bits > 64 ) then
+            call error_handler( 'BITS in STRING was greater than 64.', &
+                                char_string_too_large_error, status,   &
+                                module_name, procedure )
+            return
+        end if
```
I recently introduced `max_digits` and `overflow_bits` in stdlib_bitsets.fypp
      integer, parameter ::                   &
          max_digits = 10 ! bits_kind == int32
    !        max_digits = 20 ! bits_kind == int64

      integer(bits_kind), parameter ::     &
          overflow_bits = 2_bits_kind**30/5 ! bits_kind == int32
    !        overflow_bits = 2_bits_kind**62/5 ! bits_kind == int64

so that

                if (  digits == 10 .AND. bits > 2_bits_kind**30/5 ) &

can be written

                if (  digits == max_digits .AND. bits > overflow_bits ) &

so that overflow checks for reading and writing bitsets can be easily modified for bits_kind=int64. For some reason, this introduction has revealed that gfortran 10.2's runtime code generation is fragile. In order to get the test cases to run to completion I had to introduce the statement

    public :: max_digits, overflow_bits

in stdlib_bitsets.fypp and change

        if ( pos < 0 .OR. pos > self % num_bits-1 ) return
        self % block = ibset( self % block, pos )

in stdlib_bitsets_64.fypp to

        integer(block_kind) :: dummy
        if ( pos < 0 .OR. pos > self % num_bits-1 ) return
        dummy = ibset( self % block, pos )
        self % block = dummy

in order for the test cases to run to completion. I don't think that either change is required by the standard. Certainly other parameters used by the submodules do not have to be `public`, and the comparable code in stdlib_bitsets_large.fypp

        if ( pos < 0 .OR. pos > self % num_bits-1 ) return

        set_block = pos / block_size + 1
        block_bit = pos - (set_block - 1) * block_size
        self % blocks(set_block) = ibset( self % blocks(set_block), block_bit )

still seems to be working.

jvdp1 commented on src/stdlib_bitsets_large.fypp:
```diff
@@ -0,0 +1,1348 @@
+            case(ia0:ia9)
+                digits = digits + 1
+                if ( digits == max_digits .AND. bits > overflow_bits ) go to 996
+!! May not be quite right
```
```suggestion
!! May not be quite right
```
What does this comment mean? What could be not quite right?

jvdp1 commented on src/tests/bitsets/test_stdlib_bitset_64.f90:
```diff
@@ -0,0 +1,744 @@
+program test_stdlib_bitset_64
+    use, intrinsic :: iso_fortran_env, only : int8, int16, int32, int64
```
```suggestion
    use stdlib_kinds, only : int8, int16, int32, int64
```

jvdp1 commented on src/tests/bitsets/test_stdlib_bitset_64.f90:
```diff
@@ -0,0 +1,744 @@
+    subroutine test_string_operations()
+        character(*), parameter:: procedure = 'TEST_STRING_OPERATIONS'
+
+        write(*,*)
+        write(*,*) 'Test string operations: from_string, read_bitset, ' // &
+            'to_string, and write_bitset'
```
```suggestion
        write(*,'(/a)') 'Test string operations: from_string, read_bitset, ' // &
            'to_string, and write_bitset'
```

wclodius2 replied:
OK

jvdp1 commented on src/tests/bitsets/test_stdlib_bitset_64.f90:
```diff
@@ -0,0 +1,744 @@
+            'to_string, and write_bitset'
+
+        call set0 % from_string( bitstring_0 )
+        if ( bits(set0) /= 33 ) then
```
Suggestion only (does not need to be applied if not wanted/too much work):
```suggestion
        call check(bits(set0) /= 33, msg =  error stop procedure // ' from_string failed to interpret ' // &
                'bitstring_0 size properly.' )
```

wclodius2 replied:
Doing it everywhere seems like a lot of work, particularly for those cases with `if...then...else if...` cascades, and doing it for the few cases without the `if...then...else if...` cascades is liable to result in an inconsistent style of error reporting.

jvdp1 replied:
I understand your comment. Good for me as it is.

jvdp1 commented on src/tests/bitsets/test_stdlib_bitset_64.f90:
```diff
@@ -0,0 +1,744 @@
+
+        call set3 % read_bitset( bitstring_0, status )
+        if ( status /= success ) then
+            write(*,*) 'read_bitset_string failed with bitstring_0 as expected.'
```
Is an `error stop` not missing here?

wclodius2 replied:
I guess an error stop if the expected failure does not occur is appropriate.

jvdp1 commented on src/tests/bitsets/test_stdlib_bitset_64.f90:
```diff
@@ -0,0 +1,744 @@
+                'advance=="no" succeeded.'
+        end if
+
+        open( newunit=unit, file='test.bin', status='replace', &
```
I suggest to add some tests with `access='stream', form = 'unformatted'`

wclodius2 replied:
OK

jvdp1 commented on src/tests/bitsets/test_stdlib_bitset_64.f90:
```diff
@@ -0,0 +1,744 @@
+
+        write(*,*) 'As expected set0 had no bits set.'
+
+        go to 110
```
Is this `go to 110` really needed?

wclodius2 replied:
I don't see how to avoid using a go to here.

wclodius2 replied:
Now I do.

jvdp1 commented on src/tests/bitsets/test_stdlib_bitset_64.f90:
```diff
@@ -0,0 +1,744 @@
+
+        go to 110
+
+100     error stop procedure // ' against expectations set0 has ' // &
```
`100` is only used once Therefore, I suggest to replace the `go to 100` by the `error stop` directly in te code.

jvdp1 commented on src/tests/bitsets/test_stdlib_bitset_64.f90:
```diff
@@ -0,0 +1,744 @@
+
+        go to 210
+
+200     error stop procedure // ' against expectations set1 has ' // &
+                   'at least 1 bit unset.'
+210     continue
```
Same comment as previous: replace 200 by the error stop.

wclodius2 replied:
Done

jvdp1 commented on src/tests/bitsets/test_stdlib_bitset_large.f90:
```diff
@@ -0,0 +1,1476 @@
+            error stop procedure // ' failed to interpret bitstring_all ' // &
+                '// bitstring_all value properly.'
+        else
+            write(*,*) 'from_string transferred bitstring_all // ' // &
```
```suggestion
            write(*,*) 'from_string transferred bitstring_all ' // &
```

wclodius2 replied:
The original code is correct if odd.

jvdp1 replied:
I understand the code now. Sorry for the misunderstanding.

jvdp1 commented on src/tests/bitsets/test_stdlib_bitset_large.f90:
```diff
@@ -0,0 +1,1476 @@
+            error stop procedure // ' read_bitset_string failed to ' // &
+                'interpret "s33b" // bitstring_0 size properly.'
+        else if ( .not. set3 % none() ) then
+            error stop procedure // ' failed to interpret "s33b" // ' // &
```
```suggestion
            error stop procedure // ' failed to interpret "s33b" ' // &
```

wclodius2 replied:
Creating a string about concatenation looks odd, but is correct.

jvdp1 commented on src/tests/bitsets/test_stdlib_bitset_large.f90:
```diff
@@ -0,0 +1,1476 @@
+            error stop procedure // ' failed to interpret "s33b" // ' // &
+                'bitstring_0 value properly.'
+        else
+            write(*,*) 'read_bitset_string transferred "s33b" // ' // &
```
```suggestion
            write(*,*) 'read_bitset_string transferred "s33b" ' // &
```

jvdp1 commented on src/tests/bitsets/test_stdlib_bitset_large.f90:
```diff
@@ -0,0 +1,1476 @@
+            error stop procedure // ' read_bitset_string failed to // ' // &
+                'interpret "s33b" bitstring_all value properly.'
+        else if ( .not. set4 % all() ) then
+            error stop procedure // ' read_bitset_string failed to // ' // &
```
```suggestion
            error stop procedure // ' read_bitset_string failed to ' // &
```

jvdp1 commented on src/tests/bitsets/test_stdlib_bitset_large.f90:
```diff
@@ -0,0 +1,1476 @@
+            error stop procedure // ' read_bitset_string failed to ' // &
+                'interpret "s33b" // bitstring_all value properly.'
+        else if ( .not. set4 % any() ) then
+            error stop procedure // ' read_bitset_string failed to // ' // &
```
```suggestion
            error stop procedure // ' read_bitset_string failed to ' // &
```

jvdp1 commented on src/tests/bitsets/test_stdlib_bitset_large.f90:
```diff
@@ -0,0 +1,1476 @@
+        end if
+
+        open( newunit=unit, file='test.bin', status='replace', &
+            form='unformatted', action='write' )
```
It would be nice to have the same tests with `access=stream`

wclodius2 replied:
Done

jvdp1 replied:
Thank you.

jvdp1 commented on src/tests/bitsets/test_stdlib_bitset_large.f90:
```diff
@@ -0,0 +1,1476 @@
+
+        write(*,*) 'As expected set0 had no bits set.'
+
+        go to 110
+
+100     error stop procedure // ' against expectations set0 has ' // &
+                    'at least 1 bit set.'
+
+110     continue
```
These series of `go to` are not needed, right?

wclodius2 replied:
Yes they can be recoded and I have done so.

jvdp1 reviewed (commented):
Here are some additional comments on the tests. All options seem to be covered by the tests. I added some suggestions.
On overall I am pleased with this PR. Thank you @wclodius2!

wclodius2 commented:
It means I wrote the code when I was tired and wanted someone else to check my reasoning. It is parsing a decimal representation of the size in bits, and I want to be certain that the size does not exceed what can be represented by an integer of `bits_kind`. That could happen for example if parsing a file with `bits_kind==int32` and the file was written with `bits_kind==int64`. We want to be sure that the representation for `bits_kind==int32` does not exceed `2**31-1`. My reasoning is that it will exceed that value if:
1. it has ten digits and is about to read another digit increasing its value by a factor of ten
2. It has nine digits and has a value greater than `2**31/10`
3. It had a value of `2**31/10` and reading the next digit causes an integer overflow of nine or less resulting in a negative value.

> On Oct 20, 2020, at 11:46 AM, Jeremie Vandenplas <notifications@github.com> wrote:
> 
> 
> @jvdp1 commented on this pull request.
> 
> In src/stdlib_bitsets_large.fypp <https://github.com/fortran-lang/stdlib/pull/239#discussion_r508720956>:
> 
> > +        end do find_start
> +
> +        if ( pos > len(string) - 8 ) go to 999
> +
> +        if ( string(pos:pos) /= 's' .AND. string(pos:pos) /= 'S' ) go to 999
> +
> +        pos = pos + 1
> +        bits = 0
> +        digits = 0
> +
> +        do
> +            select case( iachar( string(pos:pos) ) )
> +            case(ia0:ia9)
> +                digits = digits + 1
> +                if ( digits == max_digits .AND. bits > overflow_bits ) go to 996
> +!! May not be quite right
> ⬇️ Suggested change
> -!! May not be quite right
> +!! May not be quite right
> What does this comment mean? What could be not quite right?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/fortran-lang/stdlib/pull/239#pullrequestreview-512984541>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/APTQDOUBL3KX6BCX26LADBTSLXEG5ANCNFSM4R6QFU5A>.
>

wclodius2 commented:
I have now resolved my qualms about my handling of potential integer overflows on reads. I have changed `max_digits` for `bits_kind==int64` from 20 to 19 (`2**63-1` is about `9.2E18`) and convinced myself that the rest of my logic was correct. I have move the comments that the code may not be quite right. (I also fixed a typo in an error message.)

wclodius2 commented:
@jvdp1 thanks for the thorough review.

jvdp1 commented on src/tests/bitsets/test_stdlib_bitset_large.f90:
```diff
@@ -0,0 +1,1504 @@
+
+        call set10 % not()
+        do i=0, set10 % bits() - 1
+            if ( set10 % test(i) ) go to 500
+        end do
+
+        write(*,*) 'As expected set10 had no bits set.'
+
+        go to 510
+
+500     error stop procedure // ' against expectations set10 has ' // &
+            'at least 1 bit set.'
+
+510     continue
+
+        do i=0, set11 % bits() - 1
+            if ( .not. set11 % test(i) ) go to 600
+        end do
+
+        write(*,*) 'As expected set11 had all bits set.'
+
+        go to 610
+
+600     error stop procedure // ' against expectations set11 has ' // &
+            'at least 1 bit unset.'
+
+610     continue
+
+        do i=0, set10 % bits() - 1
+            if ( set10 % value(i) /= 0 ) go to 700
+        end do
+
+        write(*,*) 'As expected set10 had no bits set.'
+
+        go to 710
+
+700     error stop procedure // ' against expectations set10 has ' // &
+                    'at least 1 bit set.'
+
+710     continue
```
These gotos and the following ones could be removed too, right?

wclodius2 replied:
Done. Sorry I missed them.

milancurcic commented on doc/specs/stdlib_bitsets.md:
```diff
@@ -0,0 +1,2000 @@
+
+The `stdlib_bitsets` module implements bitset types. A bitset is a
+compact representation of a sequence of `bits` binary values. It can
+equivalently be considered as a sequence of logical values or as a
+subset of the integers 0 ... `bits-1`. The bits are indexed from 0 to
```
> or as a subset of the integers 0 ... `bits-1`.

This confuses me. The equivalent integer values are 0 and 1, not 0 through `bits-1`, correct?

Do you mean to say something like this?

"It can equivalently be considered as a sequence of logical values or as a sequence of integers 0 and 1 with indices in the range 0... `bits-1`?"

wclodius2 commented:
A bit set for each index `bit` in the range `0…bits-1` has a value of 0 or 1.  A value of 1 for index `bit` can be considered as including the `bit` value in the subset, i.e., the value `1110` can be considered as defining the subset (1, 2, 3). This was how Wirth defined sets in Pascal.

> On Nov 13, 2020, at 11:42 AM, Milan Curcic <notifications@github.com> wrote:
> 
> 
> @milancurcic commented on this pull request.
> 
> In doc/specs/stdlib_bitsets.md <https://github.com/fortran-lang/stdlib/pull/239#discussion_r523154238>:
> 
> > +equivalently be considered as a sequence of logical values or as a
> +subset of the integers 0 ... `bits-1`. The bits are indexed from 0 to
> or as a subset of the integers 0 ... bits-1.
> 
> This confuses me. The equivalent integer values are 0 and 1, not 0 through bits-1, correct?
> 
> Do you mean to say something like this?
> 
> "It can equivalently be considered as a sequence of logical values or as a sequence of integers 0 and 1 with indices in the range 0... bits-1?"
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/fortran-lang/stdlib/pull/239#pullrequestreview-530317413>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/APTQDOTR4LOOIPRJ6D2QW63SPV42PANCNFSM4R6QFU5A>.
>

milancurcic commented:
Okay, thanks, I understand now. All good.

milancurcic commented:
Your example "i.e., the value `1110` can be considered as defining the subset (1, 2, 3)" was what made it click for me. Do you agree that we include it in the spec? Something like this:

"It can equivalently be considered as a sequence of logical values or as a subset of the integers 0 ... `bits-1`. For example, the value `1110` can be considered as defining the subset of integers [1, 2, 3]."

wclodius2 commented:
Sounds good to me.

milancurcic commented:
I will go ahead and merge this considering there aren't any objections. Thank you @wclodius2 and all the reviewers.
